src/boot/efi/chid.c

```diff
@@ -0,0 +1,385 @@
+                len = strlen16((char16_t *) str) * sizeof(char16_t);
+
+                if (len == 0)
+                        continue;
+
+                status = prot->HashUpdate(prot, str, len);
+                if (EFI_STATUS_IS_ERROR(status))
+                        goto exit;
+        }
+
+        status = prot->HashFinal(prot, (EFI_HASH2_OUTPUT *) hash);
+        if (EFI_STATUS_IS_ERROR(status))
+                goto exit;
+
+        status = EFI_SUCCESS;
+
+exit:
+        va_end(args);
+        return status;
+}
+
+typedef struct SmbiosInfo {
+        char16_t *manufacturer;
+        char16_t *product_name;
+        char16_t *product_sku;
+        char16_t *family;
+        char16_t *baseboard_product;
+        char16_t *baseboard_manufacturer;
+} SmbiosInfo;
+
+/**
+ * smbios_to_hashable_string() - Convert ascii smbios string to stripped char16_t.
+ */
+static char16_t *smbios_to_hashable_string(const char *str) {
+        int len, i;
+        char16_t *ret;
+
+        if (!str) {
+                /* User of this function is expected to free the result. */
+                ret = xcalloc(sizeof(*ret));
+                return ret;
+        }
+
+        /*
+         * We need to strip leading and trailing spaces, leading zeroes.
+         * See fwupd/libfwupdplugin/fu-hwids-smbios.c
+         */
+        while (*str == ' ')
+                str++;
+
+        while (*str == '0')
+                str++;
+
+        len = strlen8(str);
+
+        while (len && str[len - 1] == ' ')
+                len--;
+
+        ret = xcalloc((len + 1) * sizeof(*ret));
+        if (!ret)
+                return NULL;
+
+        for (i = 0; i < len; ++i)
+                ret[i] = str[i];
+
+        return ret;
+}
+
+static EFI_STATUS populate_smbios_info(SmbiosInfo *info) {
+        RawSmbiosInfo raw;
+        EFI_STATUS status;
+
+        status = smbios_raw_info_populate(&raw);
+        if (EFI_STATUS_IS_ERROR(status))
+                return status;
+
+        info->manufacturer = smbios_to_hashable_string(raw.manufacturer);
+        info->product_name = smbios_to_hashable_string(raw.product_name);
+        info->product_sku = smbios_to_hashable_string(raw.product_sku);
+        info->family = smbios_to_hashable_string(raw.family);
+        info->baseboard_product = smbios_to_hashable_string(raw.baseboard_product);
+        info->baseboard_manufacturer = smbios_to_hashable_string(raw.baseboard_manufacturer);
+
+        return EFI_SUCCESS;
+}
+
+static void smbios_info_clear(SmbiosInfo *info) {
+        info->manufacturer = mfree(info->manufacturer);
+        info->product_name = mfree(info->product_name);
+        info->product_sku = mfree(info->product_sku);
+        info->family = mfree(info->family);
+        info->baseboard_product = mfree(info->baseboard_product);
+        info->baseboard_manufacturer = mfree(info->baseboard_manufacturer);
+}
+
+static EFI_STATUS get_chid(SmbiosInfo *info, int id, EFI_GUID *chid) {
```

Comment:
as per coding style last argument should be called `ret_chid`?

Reply:
hmm, the 2nd argument seems to be some kind of enum, right? maybe define it as proper enum then with useful names?

Reply:
ok, i guess i grok now that doing an enum for this is a bit weird, given these don't really have a name, they are just different combos.